### PR TITLE
feat: XYNE-57426 Add copy link option for public recordings

### PR DESCRIPTION
**Pull request by shubhamdhamija-juspay**
Migration-Changes:
  - I have to create a flag to make the recording public so anyone with the link can access it

Services-Requiring-Redeployment:
  - backend
  - dashboard

## Type of Change

- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Areas Touched

- [ ] `apps/backend` (API / worker)
- [ ] `apps/dashboard`
- [ ] `apps/electron`
- [ ] `apps/xyne-claw` / `apps/xyne-claw-auth`
- [ ] `packages/shared` or another shared package
- [ ] Infra (docker, scripts, nix, CI)

---

## Zero (Sync) Changes

- [ ] No Zero changes — **skip this section**
- [ ] Adds/modifies a **mutator**
- [ ] Adds/modifies the **schema** (tables, columns, relationships)
- [ ] Adds/modifies a **query or ACL**

If any of the above:

- [ ] Server (`apps/backend/src/zero/mutators.ts`) and client (`apps/dashboard/src/zero/mutators.ts`) mutators mirror each other — same namespace, name, and args schema
- [ ] No `Date.now()` / `uuid()` generated inside a mutator
- [ ] Optimistic client result matches the server result (no state flash on rebase)
- [ ] New tables exported in `createSchema()` and given a Query ACL registered in `query-acl-factory.ts`
- [ ] Matching Prisma model + migration, client regenerated

## Schema & Data Changes

- [ ] No schema changes — **skip this section**
- [ ] Prisma schema modified (`apps/backend/prisma` and/or `prisma-common`)
- [ ] Migration added
- [ ] Backfill / data migration included

If any of the above:

- [ ] Clients regenerated (`db:generate`, `db:common:generate`)
- [ ] No new Postgres enum or enum value — enums are frozen (`pnpm run test:enum`)
- [ ] Backward compatible with deployed code, or rollout order noted below
- [ ] Backfill is idempotent

## Config, Environment & URLs

- [ ] No config changes — **skip this section**
- [ ] Env variable added / renamed / removed
- [ ] **Base URL, host, port, or origin changed** (API, Zero, Vespa, Claw, LiveKit, LiteLLM, …)
- [ ] CORS / allowed origins / OAuth redirect URIs changed
- [ ] Feature flag or runtime config changed

If any of the above:

- [ ] Key added to **all** relevant env files, not just the one you run — `apps/backend/.env.{example,local,test}`, `apps/dashboard/.env.{example,local,test}`, plus the app you touched
- [ ] Env setup / secret scripts and docker compose updated
- [ ] Verified in every consumer with its own base URL — dashboard (`VITE_API_BASE_URL`, `VITE_ZERO_SERVER`), Electron (`VITE_ELECTRON_*`), shareable links (`VITE_SHAREABLE_ORIGIN`), server-to-server (`BACKEND_URL`, `XYNE_CLAW_URL`, `XYNE_CLAW_AUTH_URL`)
- [ ] Google / Microsoft OAuth redirect URIs still resolve
- [ ] Nothing hardcoded, no secrets committed

## API Contract

- [ ] No API changes
- [ ] New endpoint
- [ ] Existing contract modified (shape, status codes, auth)
- [ ] Breaking for dashboard / electron / external / bots — migration noted above

---

### Automated

- [ ] Backend unit tests — `pnpm --filter xyne-spaces-backend run test`
- [ ] Claw tests — `pnpm --filter xyne-claw run test`
- [ ] End-to-end suite — `pnpm run test`
- [ ] Added / updated tests for this change
- [ ] Not covered by tests 

### Manual

**Users**
- [ ] Single user
- [ ] Two users at once (two accounts / two browsers) — sync lands on both, no cross-user leakage
- [ ] Different roles or permission levels (member vs admin, ACL boundaries)
- [ ] Different workspaces / spaces — data stays scoped

**Login**
- [ ] Google
- [ ] Microsoft
- [ ] Dev auth (`ENABLE_DEV_AUTH`)
- [ ] Fresh signup / first-time user
- [ ] Session expiry, re-login, logout

**Run mode**
- [ ] Local dev (`pnpm run dev:all`)
- [ ] Test mode (`dev:test`)
- [ ] Sandbox / claw test (`claw:test`)
- [ ] Prod-like local (`dev:prod`)
- [ ] Docker compose
- [ ] Electron desktop

**Sync & resilience** — for anything touching Zero
- [ ] Multiple tabs stay consistent
- [ ] Reload / hard refresh — no duplicate or lost rows
- [ ] Offline then reconnect — pending mutations replay cleanly
- [ ] Concurrent edits on the same entity by two users

**Surface & data**
- [ ] Web browser
- [ ] Electron desktop
- [ ] Narrow viewport, dark + light theme
- [ ] Empty state
- [ ] Large / realistic data volume
- [ ] Error paths (network failure, denied permission, invalid input)

---

## Checklist

- [ ] Reviewed my own diff; scoped to one thing
- [ ] `pnpm run build:shared` passes
- [ ] Backend `typecheck` + `build` pass
- [ ] Dashboard `lint:errors-only` + `typecheck` + `build` pass
- [ ] Formatting clean in the packages I touched
- [ ] New deps added with `pnpm --filter <pkg> add <dep>`; version pins in root `pnpm.overrides`
- [ ] Commits follow `<type>: <TICKET-ID> <subject>`; branch is `fix/*` or `feature/*`
- [ ] Docs / guidelines updated where behaviour changed
- [ ] No debug logging or commented-out code left behind

### CODE REVIEW

**Review by XyneSpaces — COMMENTED** [review]
Xyne Spaces Reviewer inline findings for PR #1011.

File: apps/backend/prisma/migrations/20260825120000_access_policy_for_entity_access/migration.sql [hunk 1]
```diff
@@ -0,0 +1,9 @@
+-- AlterTable
+ALTER TABLE "public"."entity_access" ADD COLUMN "accessPolicy" TEXT;
+
+-- At most one workspace-public (target-less) row per entity. The model's
+-- compound unique treats NULL userId as non-conflicting, so it cannot guard
+-- this row shape.
+CREATE UNIQUE INDEX "entity_access_public_row_key"
```
Comment (XyneSpaces) [inline apps/backend/prisma/migrations/20260825120000_access_policy_for_entity_access/migration.sql]: **[blocking] R3 / 🔴 DANGEROUS public.entity_access** — `CREATE UNIQUE INDEX` is not `CONCURRENTLY` and this Prisma migration has multiple statements, so on a populated `entity_access` table it takes a write-blocking lock for the full index build and can abort if duplicate target-less rows already exist. ↳ FIX: ship the unique index concurrently in its own non-transactional/off-peak step (or otherwise use the repo's safe index rollout path), with a duplicate pre-check/backfill before enforcing uniqueness.

File: apps/backend/src/services/recordingSharingService.ts [hunk 2]
```diff
@@ -153,9 +159,92 @@ export class RecordingSharingService {
         return this.linkTicket(callId, actor, command.ticketId);
       case 'unlink_ticket':
         return this.unlinkTicket(callId, actor);
+      case 'set_public':
```
Comment (XyneSpaces) [inline apps/backend/src/services/recordingSharingService.ts]: **[major] R7** — This adds the new `set_public` sharing path plus ACL-visible workspace-wide access, but the PR does not add/update tests for creator-only toggling, public-view-only access, unpublish behavior, and cross-user/workspace visibility. Add backend service/controller tests and a small dashboard/Zero regression so this access-control behavior is covered before merge.

**Comment by XyneSpaces** [comment]
## Xyne Spaces Reviewer — PR #1011

**Head reviewed:** `3418419fd67765e4b71c9eb4abb261556b3cd933`
**Verdict:** Blocking issues found
**Reviewed:** 14 files, 17 changed-file entries/hunks · **Findings:** 1 blocking, 1 major, 0 nits

### Findings
| Severity | Rule | Location | Issue |
|----------|------|----------|-------|
| blocking | R3 / Migration Safety | apps/backend/prisma/migrations/20260825120000_access_policy_for_entity_access/migration.sql:7 | Non-`CONCURRENTLY` `CREATE UNIQUE INDEX` on the existing `public.entity_access` table can write-stall the table for the whole index build and can abort if duplicate target-less rows already exist. |
| major | R7 | apps/backend/src/services/recordingSharingService.ts:162 | New public-recording sharing/access-control path has no added or updated tests for creator-only toggling, public-view-only access, unpublish behavior, or cross-user/workspace visibility. |

### 🛢️ Migration Safety — unsafe operations found
cc @qa-xyne

| Migration | Tier | Statement (schema.table — op) | What it breaks | Fix |
|-----------|------|-------------------------------|----------------|-----|
| 20260825120000_access_policy_for_entity_access | 🔴 DANGEROUS | public.entity_access — CREATE UNIQUE INDEX | `CREATE UNIQUE INDEX` takes a lock that blocks writes for the index build on a populated table and aborts if duplicate target-less rows already exist. | Use a safe unique-index rollout: duplicate pre-check/backfill, then concurrent/off-peak index creation in its own migration/runbook path. |

Deploy stance for QA: expect access-policy visibility to affect recording/canvas access; verify publish/unpublish from web and that existing shared viewers/workspace members behave correctly after deploy.

### Recommended reviewer
**@sayantanp-eng** (request review attempted; API tool unavailable in this run) — strongest changed-file familiarity found for the recording sharing UI path, and no pending requested-review workload was found from the available open-PR data.

| Candidate | Owns (files in this PR) | Default reviewer? | Pending reviews | Tie-breaker |
|-----------|-------------------------|-------------------|-----------------|-------------|
| @sayantanp-eng | RecordingShareModal.tsx / recording actions area (recent PR #909 touched the same sharing surface) | no CODEOWNERS found | 0 found from available open-PR reviewer data | Maps to Spaces `sayantan.paul@juspay.in` |

Rationale: no CODEOWNERS file was present, so the pick falls back to the strongest file-owner signal for this recording-sharing surface plus taggable Spaces mapping.

File: apps/backend/src/services/recordingSharingService.ts (hunk 2 above)
Comment (shubhamdhamija-juspay) [reply apps/backend/src/services/recordingSharingService.ts]: No tests required

File: apps/backend/prisma/migrations/20260825120000_access_policy_for_entity_access/migration.sql (hunk 1 above)
Comment (shubhamdhamija-juspay) [reply apps/backend/prisma/migrations/20260825120000_access_policy_for_entity_access/migration.sql]: Removed

File: apps/backend/prisma/generated/zero/schema.ts [hunk 3]
```diff
@@ -1692,6 +1692,7 @@ export const entityAccessTable = table("entity_access")
     userGroupId: string().optional(),
     channelId: string().optional(),
     entityUserAccess: string(),
+    accessPolicy: string().optional(),
```
Comment (Mayankbansal-juspay) [inline apps/backend/prisma/generated/zero/schema.ts]: visibility would make for sense as it's used in most cases

File: apps/backend/prisma/migrations/20260825120000_access_policy_for_entity_access/migration.sql [hunk 4]
```diff
@@ -0,0 +1,2 @@
+-- AlterTable
+ALTER TABLE "public"."entity_access" ADD COLUMN "accessPolicy" TEXT;
```
Comment (Mayankbansal-juspay) [inline apps/backend/prisma/migrations/20260825120000_access_policy_for_entity_access/migration.sql]: is this auto-generated or llm generated

File: apps/backend/src/database/acl/tables/calls-acl.ts [hunk 5]
```diff
@@ -41,6 +41,14 @@ export class CallsACL extends BaseQueryACL<
           { userId: this.ctx.userId },
           ...(userGroupIds.length ? [{ userGroupId: { in: userGroupIds } }] : []),
           ...(channelIds.length ? [{ channelId: { in: channelIds } }] : []),
+          // Workspace-public: target-less row with accessPolicy PUBLIC (full
+          // members only — guests never reach this branch).
```
Comment (Mayankbansal-juspay) [inline apps/backend/src/database/acl/tables/calls-acl.ts]: can we remove this inline comment

File: apps/backend/src/database/acl/tables/calls-acl.ts [hunk 6]
```diff
@@ -41,6 +41,14 @@ export class CallsACL extends BaseQueryACL<
...
+          {
+            accessPolicy: EntityAccessPolicy.PUBLIC,
+            userId: null,
```
Comment (Mayankbansal-juspay) [inline apps/backend/src/database/acl/tables/calls-acl.ts]: these explicit nulls are not needed ig

File: apps/backend/src/database/acl/tables/calls-acl.ts [hunk 7]
```diff
@@ -41,6 +41,14 @@ export class CallsACL extends BaseQueryACL<
           { userId: this.ctx.userId },
           ...(userGroupIds.length ? [{ userGroupId: { in: userGroupIds } }] : []),
           ...(channelIds.length ? [{ channelId: { in: channelIds } }] : []),
+          // Workspace-public: target-less row with accessPolicy PUBLIC (full
+          // members only — guests never reach this branch).
+          {
+            accessPolicy: EntityAccessPolicy.PUBLIC,
```
Comment (Mayankbansal-juspay) [inline apps/backend/src/database/acl/tables/calls-acl.ts]: can we make this public as the first step, it will reduce the query time

File: apps/backend/src/services/callShareService.ts [hunk 8]
```diff
@@ -9,12 +9,13 @@ import { ShareableEntityType } from '@xyne/shared';
...
   async canView(call: Call, userId: string, workspaceId: string): Promise<boolean> {
     if (call.createdByUserId === userId) return true;
-    return entityAccessService.hasActiveShare({
+    const key = {
       workspaceId,
       shareableEntityType: ShareableEntityType.NOTE_TAKER,
       entityId: call.id,
-      userId,
-    });
+    };
```
Comment (Mayankbansal-juspay) [inline apps/backend/src/services/callShareService.ts]: here we are passing key in hasActiveShare also and then we are explicitly calling hasPublicAccess, in hasActiveShare only we can add a check if it's public just return true

File: apps/backend/src/services/entityAccessService.ts [hunk 9]
```diff
@@ -63,6 +64,31 @@ export class EntityAccessService {
     return Boolean(viaMembership);
   }
 
+  /**
+   * Workspace-wide "general access": true when the entity has a target-less row
+   * (userId/userGroupId/channelId all NULL) with accessPolicy PUBLIC. Kept
+   * separate from hasActiveShare on purpose — targeted shares also confer
+   * manage rights on recordings, public access must only confer viewing.
+   */
```
Comment (Mayankbansal-juspay) [inline apps/backend/src/services/entityAccessService.ts]: no need of this.

File: apps/backend/src/services/recordingSharingService.ts [hunk 10]
```diff
@@ -153,9 +159,93 @@ export class RecordingSharingService {
...
+        // Never create a row just to record PRIVATE — absence already means it.
+        await tx.entityAccess.create({
+          data: {
+            id: randomUUID(),
+            workspaceId: actor.workspaceId,
+            shareableEntityType: ShareableEntityType.NOTE_TAKER,
+            entityId: recording.id,
+            entityUserAccess: EntityUserAccess.VIEW,
+            accessPolicy: policy,
+            metadata: { intent: RECORDING_SHARE_INTENT.PUBLIC_LINK },
```
Comment (Mayankbansal-juspay) [inline apps/backend/src/services/recordingSharingService.ts]: this metadata update won't be needed , as row itself will tell us the intent

File: apps/backend/src/services/recordingSharingService.ts [hunk 11]
```diff
@@ -153,9 +159,93 @@ export class RecordingSharingService {
...
+   * Toggles workspace-wide "general access" for a recording via a single
+   * target-less entity_access row (userId/userGroupId/channelId all NULL) whose
+   * `accessPolicy` is PUBLIC or PRIVATE. The row is kept on un-publish so the
+   * audit trail survives, mirroring how revoke keeps REVOKED rows.
+   *
+   * The recording's canvases follow the same toggle via syncCanvasVisibility,
+   * so a public recording's notes/summary open for workspace members and close
+   * again when access is restricted.
+   *
+   * Creator-only: unlike grant/revoke (which any active share holder may use),
```
Comment (Mayankbansal-juspay) [inline apps/backend/src/services/recordingSharingService.ts]: is it on purpose, anyone who got the access from owner can't make it public?

File: apps/backend/src/services/recordingSharingService.ts [hunk 12]
```diff
@@ -153,9 +159,93 @@ export class RecordingSharingService {
...
+   * opening a recording to the entire workspace is reserved for its creator.
+   */
+  private async setPublic(
```
Comment (Mayankbansal-juspay) [inline apps/backend/src/services/recordingSharingService.ts]: only setPublic funtion no set private function?

File: apps/backend/prisma/migrations/20260825120000_access_policy_for_entity_access/migration.sql (hunk 4 above)
Comment (Mayankbansal-juspay) [reply apps/backend/prisma/migrations/20260825120000_access_policy_for_entity_access/migration.sql]: go to main delete migrations and then generate migration for main, switch to your branch then migrate again to generate file

File: apps/backend/prisma/generated/zero/schema.ts (hunk 3 above)
Comment (shubhamdhamija-juspay) [reply apps/backend/prisma/generated/zero/schema.ts]: Done

File: apps/backend/src/database/acl/tables/calls-acl.ts (hunk 5 above)
Comment (shubhamdhamija-juspay) [reply apps/backend/src/database/acl/tables/calls-acl.ts]: removed

File: apps/backend/src/database/acl/tables/calls-acl.ts (hunk 7 above)
Comment (shubhamdhamija-juspay) [reply apps/backend/src/database/acl/tables/calls-acl.ts]: Done

File: apps/backend/src/database/acl/tables/calls-acl.ts (hunk 6 above)
Comment (shubhamdhamija-juspay) [reply apps/backend/src/database/acl/tables/calls-acl.ts]: Done

File: apps/backend/src/services/callShareService.ts (hunk 8 above)
Comment (shubhamdhamija-juspay) [reply apps/backend/src/services/callShareService.ts]: done

File: apps/backend/src/services/entityAccessService.ts (hunk 9 above)
Comment (shubhamdhamija-juspay) [reply apps/backend/src/services/entityAccessService.ts]: Removed

File: apps/backend/src/services/entityAccessService.ts [hunk 13]
```diff
@@ -36,6 +43,22 @@ export class EntityAccessService {
       return true;
     }
 
+    // Workspace-public link access grants viewing to every full member, so it
+    // short-circuits ahead of the costlier membership lookups. PUBLIC is only
+    // ever written on the target-less row, so the visibility check alone
+    // identifies it (mirrors CallsACL.getSharedCallIds).
+    const publicRow = await this.db.entityAccess.findFirst({
```
Comment (Mayankbansal-juspay) [inline apps/backend/src/services/entityAccessService.ts]: move to entityAccessRepository

File: apps/backend/src/database/acl/tables/calls-acl.ts [hunk 14]
```diff
@@ -38,6 +38,9 @@ export class CallsACL extends BaseQueryACL<
         shareableEntityType: ShareableEntityType.NOTE_TAKER,
         entityUserAccess: { not: EntityUserAccess.REVOKED },
         OR: [
+          // PUBLIC is only ever written on the target-less link-access row, so
```
Comment (Mayankbansal-juspay) [inline apps/backend/src/database/acl/tables/calls-acl.ts]: remove comment

File: apps/backend/src/controllers/recordingSharingController.ts [hunk 15]
```diff
@@ -32,6 +32,10 @@ const RecordingSharingCommandSchema = z.discriminatedUnion('action', [
     ticketId: z.string().min(1),
   }),
   z.object({ action: z.literal('unlink_ticket') }),
+  z.object({
+    action: z.literal('set_public'),
```
Comment (Mayankbansal-juspay) [inline apps/backend/src/controllers/recordingSharingController.ts]: misleading

File: apps/dashboard/src/routes/RecordingDetailV2Screen/components/RecordingShareModal.tsx [hunk 16]
```diff
@@ -65,11 +69,60 @@ export const RecordingShareModal: React.FC<RecordingShareModalProps> = ({
...
+          !locallyRevokedShareIds.has(share.id) &&
+          !isRecordingTicketLinkShare(share.metadata) &&
+          !isRecordingGeneralAccessShare(share),
       ),
     [locallyRevokedShareIds, recordingRow],
   );
 
+  // Workspace-wide general access, held on the single target-less share row.
+  const isOwner = currentUser?.id === recording.createdByUserId;
+  const serverIsPublic = useMemo(
```
Comment (Mayankbansal-juspay) [inline apps/dashboard/src/routes/RecordingDetailV2Screen/components/RecordingShareModal.tsx]: check

File: apps/backend/src/controllers/recordingSharingController.ts (hunk 15 above)
Comment (shubhamdhamija-juspay) [reply apps/backend/src/controllers/recordingSharingController.ts]: Done

File: apps/backend/src/database/acl/tables/calls-acl.ts (hunk 14 above)
Comment (shubhamdhamija-juspay) [reply apps/backend/src/database/acl/tables/calls-acl.ts]: Removec

File: apps/backend/src/services/entityAccessService.ts (hunk 13 above)
Comment (shubhamdhamija-juspay) [reply apps/backend/src/services/entityAccessService.ts]: done

File: apps/dashboard/src/utils/recordingUtils.ts [hunk 17]
```diff
@@ -234,6 +234,17 @@ export const isRecordingTicketLinkShare = (metadata: unknown): boolean => {
   return (metadata as Record<string, unknown>)['intent'] === 'ticket_link';
 };
 
+/**
+ * The workspace-wide "general access" row: no user, group, or channel target.
+ * Keyed on the null targets (not metadata/visibility) so a target-less row
+ * can never be rendered as a person in the share list, whatever its state.
+ */
+export const isRecordingGeneralAccessShare = (share: {
```
Comment (Mayankbansal-juspay) [inline apps/dashboard/src/utils/recordingUtils.ts]: check
